### ISSUE TITLE
Label entities by the component bundles they have for clearer inspection

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#1958]
## What problem does this solve or what need does it fill?

When inspecting entities for debugging or as part of an editor, it can often be challenging to quickly get a a sense of what "kind" of entity it is. 

Right now, you can either inspect it visually, or pore through its (long) list of components. This will become unwieldy when we need things like tooltip displays.

## What solution would you like?

Create "entity labels", which are many-to-many and use the same type magic as e.g. system and stage labels.

These can be added to the AppBuilder with the `label_entities(bundle: impl Bundle)`.

Each entity will have a "labels" field, which will show all relevant labels in alphabetical order (there should only be one in most cases). A label is relevant to an entity if it has all of the components in that bundle.

These will be almost entirely for display purposes. Under the hood, this can be computed and stored on a per archetype level.

Depending on their cost, we may want to only enable these during debug builds.

## What alternative(s) have you considered?

Somehow designate a specific component as being "defining", or specify a priority of these.

## Additional context

These will be used in close association with archetype invariants (#1481). We should ensure that the syntax and location of these are closely tied.

These bundles will commonly be reused for kinded entities; it may be correct to allow syntactic sugar to allow their specification via these entity labels.

In theory we could use query filters to search for these entities. I'm not entirely convinced by this plan, but it's not the worst idea.

**Comment by mockersf** [comment]
Using strings for label, we have the [`Labels`](https://docs.rs/bevy/0.5.0/bevy/core/struct.Labels.html) component to be able to add labels to an entity, and the [`EntityLabels`](https://docs.rs/bevy/0.5.0/bevy/core/struct.EntityLabels.html) to query entities by label

**Comment by cart** [comment]
> Create "entity labels", which are many-to-many and use the same type magic as e.g. system and stage labels.
> These will be almost entirely for display purposes. Under the hood, this can be computed and stored on a per archetype level.

Functionally this sounds identical to zero-sized marker components, which are cheap, queryable, integrate with scenes, and already have all relevant CRUD ops defined. It sounds like this is largely a categorization issue? Ex: maybe we need a way to promote specific component types to "labels" for things like editors?

**Comment by cart** [comment]
> Somehow designate a specific component as being "defining", or specify a priority of these.

lol you already covered this. i guess i glazed over it :smile:

**Comment by cart** [comment]
Maybe we could just track "bundle components" somewhere? Ex: we could just add `is_bundle_component` to `ComponentInfo`. I think I'd prefer something like that to creating a brand new "typed component label" concept because it builds on what is already there and doesn't require increasing the user-facing api surface.

**Comment by alice-i-cecile** [comment]
> maybe we need a way to promote specific component types to "labels" for things like editors?

Yeah, I'm leaning towards this as my preferred flavor of solution. I'm definitely in favor of trying to integrate this smoothly with minimal API changes. This isn't urgent or very impactful, but it could be a nice QoL feature once we have an editor.

> Maybe we could just track "bundle components" somewhere?

What if we had an `Identifier` trait that we could add to components? Or a `#[identifier]` marker that you could stick onto your fields in a bundle like we do with `#[bundle]`?

**Comment by cart** [comment]
> What if we had an Identifier trait that we could add to components? Or a #[identifier] marker that you could stick onto your fields in a bundle like we do with #[bundle]?

My problem with this is that it adds a new "Identifier" concept, which I'm not convinced we need. Bundles seem like a pretty reasonable "identifier" as-is. They denote that the entity was created for a specific purpose. Yes, there can be multiple bundles per entity, but we'd almost certainly want to support multiple "identifiers" anyway.

**Comment by alice-i-cecile** [comment]
> we'd almost certainly want to support multiple "identifiers" anyway.
Definitely agree.

> Bundles seem like a pretty reasonable "identifier" as-is. They denote that the entity was created for a specific purpose.

I think that's fair, but I'm not sure what information you'd extract from them to display. The bundle name, and then all of your identifiers are `*Bundle`? I guess we could strip that off automatically actually...

**Comment by alice-i-cecile** [comment]
I think you'd want to track this at the entity level, and set a special "bundle labels" field or something when you use `.spawn_bundle` or `.insert_bundle`

**Comment by cart** [comment]
> I think you'd want to track this at the entity level, and set a special "bundle labels" field or something when you use .spawn_bundle or .insert_bundle

Yeah I think this should be entity level. If someone _happens_ to spawn the same components a bundle has, I don't think that should be "auto labeled" with that bundle. I personally think inserting a special "bundle component" is the best way to do this.

**Comment by concave-sphere** [comment]
> I think that's fair, but I'm not sure what information you'd extract from them to display. The bundle name, and then all of your identifiers are `*Bundle`? I guess we could strip that off automatically actually...

That brings up something I've wondered about: why do bundles have to be named `*Bundle`?  I feel like for most of the bundles I create, either it's clear from context, or they're in a well labeled module (like `foo::entities::Blah`).  And it gets even more redundant when I write `commands.spawn_bundle(foo::entities::BlahBundle)` -- I just wrote 3 times on that line that it's a bundle.  It seems like `commands.spawn_bundle(foo::Blah)` would have gotten the point across clearly enough.

But I still feel the need to add the suffix.  I think this might be a subconscious attempt to avoid accidentally using a `Bundle` as a `Component`.  Maybe the suffix could be dropped if `Component` was opt-in (https://github.com/bevyengine/bevy/issues/1843)?

---

That aside: I like the idea of using `Bundle` to label things.  It makes a `Bundle` more meaningful than just a pile of components that happen to go together, and means that users get more features automatically by using the system the way it was designed.

**Comment by cart** [comment]
> why do bundles have to be named *Bundle? 

Its largely an issue of namespacing. If we have a `Sprite` component and a `Sprite` bundle, it makes using those types confusing and error prone. As a general principle, I try to treat Bevy as a global namespace (even though this isn't literally true). This ensures that types can be used side-by-side (and discussed in documentation and tutorials) without needing to specify `bevy_sprite::component::Sprite` vs `bevy_sprite::bundle::Sprite`. With a "global namespace" `Sprite` is always just one thing (in this case: a component). 

Theres also the secondary issue of "using bundles as components and vice-versa", which might be solved by #1843, but even with that I'm still very much in favor of the *Bundle suffix.

>  It makes a Bundle more meaningful than just a pile of components that happen to go together, and means that users get more features automatically by using the system the way it was designed.

Agreed. I've been planning on using Bundles for a number of things for awhile now:

1. Better scene editor workflows (ex:all registered bundles show up in visual scene editors (like Godot Nodes) and have friendly icons to help visually indicate their purpose)
2. Using bundles in scenes to make them resilient to new component additions and removals (ex: when a new component is added to a Bundle, old scenes implicitly pick it up).

**Comment by concave-sphere** [comment]
> Its largely an issue of namespacing. If we have a Sprite component and a Sprite bundle, it makes using those types confusing and error prone.

Hmm, I'd be interested in discussing this further, but I don't want to derail this bug any more than I already have.  And it may be partly my unfamiliarity with Rust community practices showing here: I've read through the [API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html), but they are pretty sparse compared to, eg, what I'm familiar with in Go land.  On this question, for instance, there's is very complete guidance in a widely cited [blog post](https://blog.golang.org/package-names) and in Effective Go [docs](https://golang.org/doc/effective_go#names).

Is there a better forum to continue this discussion, or any community docs that you think would be a helpful reference?  (or if you don't think it's worth the time, I'll drop it)

**Comment by alice-i-cecile** [comment]
@concave-sphere #1247 would be an appropriate place to discuss this, or a seperate issue would be fine if you'd prefer :)

**Comment by alice-i-cecile** [comment]
>  So, my proposal would be to have a layered priority system for what gets displayed, call it InspectorDisplayName:
>
> 1. Name is used if present
> 2. components marked as "defining" get used if Name is not present, combining as Camera/Player etc in alphabetical order if multiple are present
> 3. entities without any name are Entity, with a color or other identifier based on the archetype hash

From [recent discussion](https://discord.com/channels/691052431525675048/789987800014651412/1390093974970306692).

> [15:23]eugineerd: Why not use require(Name::new(stringify!(SomeComponent))) for such components?
> [15:24]Alice 🌹[BEVY]
> : Hmm 🤔 I expect that would be a lot more expensive, especially for something like Mesh
> [15:24]Alice 🌹[BEVY]
> : Since then you need O(n) copies, rather than O(1)
> [15:25]Alice 🌹[BEVY]
> : I do like the elegance though
> [15:25]eugineerd: wouldn't that be a compile time constant?
> [15:26]eugineerd: I'm not sure how Name is currently stored
> [15:26]Alice 🌹[BEVY]
> : You'd still need to add the component for each
> [15:27]Alice 🌹[BEVY]
> : It's a fancy string, using https://github.com/bevyengine/atomicow/blob/main/README.md in each Name component on each named entity

Another proposed approach by @viridia:

> [15:30]DreamerTalin: Well another solution is to provide a pluggable mapper
> [15:31]DreamerTalin: I.e. given an EntityWorldRef, tell me it's name. If you don't know, then the next mapper in the chain gets a shot at it.

**Comment by alice-i-cecile** [comment]
I took a crack at this again, while working on #15636, which you can examine in https://github.com/alice-i-cecile/feathers_inspector/commit/bf7d21157f7e9e2b7badc4477dd3154ff1461477

Things I liked:

- a standardized name resolution solution that can be used by various inspectors and debug tools is great
- a priority system gives us a nice way to avoid labelling entities with both `Camera` and `Camera3d`, or allows user-defined components like `Enemy` to take priority over engine-level ones like `Mesh3d`
- trivially user extensible: no special-casing of the overrides [like in `bevy-inspector-egui`](https://github.com/jakobhellermann/bevy-inspector-egui/blob/2e9f7ebe804cd4f78b74c74b4570fedd53948e13/crates/bevy-inspector-egui/src/utils.rs#L15)
- in theory, no special registration required, but see below

Things that did not work:

- actually hooking this up via reflection requires that the trait be object-safe, see the [reflect_trait docs](https://dev-docs.bevy.org/bevy/prelude/attr.reflect_trait.html). This 
   - blocks a trait-based priority system
- even without prioritization,  `NameDefining: Component` bounds cannot be added nicely, since `Component` requires specifying which storage type is used. Boo!

Other ideas:

- use a manual `App::register_name_defining_component(priority: i8)` approach. Not *that* bad because this is rare, but still not ideal
- copy notes from [bevy_inspector_egui::InspectorOptions](https://github.com/jakobhellermann/bevy-inspector-egui/blob/93fe429ba0570405094370f31d0530c1a8f0e28d/crates/bevy-inspector-egui/src/inspector_options/mod.rs#L50) and steal however they got this to work
  - this makes a lot of sense to me: this is fundamentally "inspector-specific component metadata" and should be grouped together

**Comment by viridia** [comment]
@alice-i-cecile That's interesting. It's very different from the approach I took, but it's not bad.

First, some bikeshedding: alternate names for this trait might be:
* Denominative
* Namesake
* Titular
* Synecdoche

These are all words that mean "a property or component that gives its name to the whole".

The approach I used was rather simpler and significantly more brute force:
```rust
app.add_name_guesser(|entity: &EntityRef| {
    if entity.contains::<Camera>() {
        Some("Camera")
    } else if entity.contains::<DirectionalLight>() {
        Some("DirectionalLight")
   // (much detail omitted here)
    } else {
        None
    }
}, /* priority */ 0);
```
Not particularly efficient since it linearly searches for components, but I didn't think it needed to be. In general, I think your approach is better, however my approach does permit some additional flexibility that might not be necessary: it allows you to calculate the name based on the presence of 2 or more components, and the resulting string doesn't have to be the same as the component name.
